runtime/fundamentals/testing.md

```diff
@@ -0,0 +1,757 @@
+}
+```
+
+Then running `deno test` in the same directory tree as the configuration file
+will take these options into account.
+
+## Test definition filtering
+
+Deno provides two options for filtering tests within the test definitions
+themselves: ignoring tests and focusing on specific tests.
+
+### Filtering Out (Ignoring Tests)
+
+You can ignore certain tests based on specific conditions using the `ignore`
+boolean in the test definition. If `ignore` is set to `true`, the test will be
+skipped. This is useful, for example, if you only want a test to run on a
+specific operating system.
+
+```ts
+Deno.test({
+  name: "do macOS feature",
+  ignore: Deno.build.os !== "darwin", // This test will be ignored if not running on macOS
+  fn() {
+    // do MacOS feature here
+  },
+});
+```
+
+### Filtering In (Only Running Specific Tests)
+
+If you want to focus on a particular test and ignore the rest, you can use the
+`only` option. This tells the test framework to run only the tests with only set
+to true. Multiple tests can have this option set. However, if any test is
+flagged with only, the overall test run will always fail, as this is intended to
+be a temporary measure for debugging.
+
+```ts
+Deno.test({
+  name: "Focus on this test only",
+  only: true, // Only this test will run
+  fn() {
+    // test complicated stuff here
+  },
+});
+```
+
+## Failing fast
+
+If you have a long-running test suite and wish for it to stop on the first
+failure, you can specify the `--fail-fast` flag when running the suite.
+
+```shell
+deno test --fail-fast
+```
+
+## Reporters
+
+Deno includes three built-in reporters to format test output:
+
+- `pretty` (default): Provides a detailed and readable output.
+- `dot`: Offers a concise output, useful for quickly seeing test results.
+- `junit`: Produces output in JUnit XML format, which is useful for integrating
+  with CI/CD tools.
+
+You can specify which reporter to use with the --reporter flag:
+
+```sh
+# Use the default pretty reporter
+deno test
+
+# Use the dot reporter for concise output
+deno test --reporter=dot
+
+# Use the JUnit reporter
+deno test --reporter=junit
+```
+
+Additionally, you can write the JUnit report to a file while still getting
+human-readable output in the terminal by using the `--junit-path` flag:
+
+```sh
+deno test --junit-path=./report.xml
+```
+
+## Spying, mocking (test doubles), stubbing and faking time
+
+The [Deno standard library](/runtime/fundamentals/standard_library/) provides a
+set of functions to help you write tests that involve spying, mocking, and
+stubbing. Check out the [documentation on JSR](https://jsr.io/@std/testing) for
+more information on each of these utilities.
+
+## Assertions
+
+To help developers write tests, the
+[Deno standard library](/runtime/fundamentals/standard_library/) comes with a
+built-in assertions module which can be imported from `jsr:@std/assert`. Check
+out the [documentation on JSR for its usage](https://jsr.io/@std/assert).
```

Comment:
We already showed examples with the `assert` and `expect` API earlier in the document. Wondering if this section can be removed.